bridge/sam/pool.go

```diff
@@ -0,0 +1,377 @@
+package sam
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/0xPolygon/polygon-edge/helper/common"
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+type pool struct {
+	// write-lock is called only during the validator change process
+	// otherwise read-lock is called
+	// Changing validators occurs rarely (once per epoch)
+	changeValidatorsLock sync.RWMutex
+	validators           []types.Address
+	signatureThreshold   uint64 // required number of signatures for ready
+
+	// storage for message body
+	messageMap sync.Map // types.Hash -> []byte
+
+	// map from message Hash -> bool for message flags
+	consumedMap sync.Map
+	readyMap    sync.Map
+
+	// signatures
+	messageSignatures *messageSignaturesStore
+}
+
+func NewPool(validators []types.Address, threshold uint64) Pool {
+	return &pool{
+		changeValidatorsLock: sync.RWMutex{},
+		validators:           validators,
+		signatureThreshold:   threshold,
+		messageMap:           sync.Map{},
+		consumedMap:          sync.Map{},
+		readyMap:             sync.Map{},
+		messageSignatures:    newMessageSignaturesStore(),
+	}
+}
+
+// Add adds new message with the signature to pool
+func (p *pool) AddMessage(message *Message) {
+	p.changeValidatorsLock.RLock()
+	defer p.changeValidatorsLock.RUnlock()
+
+	if p.hasConsumed(message.Hash) {
+		// we do no longer put the signature if the message has been consumed
+		return
+	}
+
+	p.setMessageBody(message.Hash, message.Body)
+	p.tryToPromote(message.Hash)
+}
+
+// Add adds new message with the signature to pool
+func (p *pool) AddSignature(signature *MessageSignature) {
+	p.changeValidatorsLock.RLock()
+	defer p.changeValidatorsLock.RUnlock()
+
+	if p.hasConsumed(signature.Hash) {
+		// we no longer put the signature if the message has been consumed
+		return
+	}
+
+	p.messageSignatures.Put(signature)
+	p.tryToPromote(signature.Hash)
+}
+
+// ConsumeMessage sets the consumed flag and delete the message from pool
+func (p *pool) ConsumeMessage(hash types.Hash) {
+	p.consumedMap.Store(hash, true)
+
+	p.messageSignatures.RemoveMessage(hash)
+	p.messageMap.Delete(hash)
+	p.readyMap.Delete(hash)
+}
+
+// knows returns the flag indicating the message is known
+func (p *pool) isMessageKnown(hash types.Hash) bool {
+	_, ok := p.messageMap.Load(hash)
+	if !ok {
+		return false
+	}
+
+	return ok
+}
```

Comment:
Why not just return `ok` ?

Reply:
Good catch. Fixed at 95405e5209abfed5ccbad58db15f2550f6d04bf5